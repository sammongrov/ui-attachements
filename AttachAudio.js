import React, { Component } from 'react';
import { Actions } from 'react-native-router-flux';
import { Text, View, TouchableOpacity, Platform, StyleSheet } from 'react-native';
import { Recorder } from 'react-native-audio-toolkit';
import { styles } from 'react-native-theme';
import { Colors } from '@ui/theme_default';
import { Icon /* , FontAwesomeIcon */ } from '@ui/components';

// TODO: clean old styles of audio attachment and add new to the theme
const customStyles = StyleSheet.create({
  audioAttachContainer: {
    paddingBottom: Platform.select({
      ios: 35,
      android: null,
    }),
  },
  composer: {
    marginLeft: 10,
    marginTop: Platform.select({
      ios: 6,
      android: 0,
    }),
    marginBottom: Platform.select({
      ios: 5,
      android: 3,
    }),
    borderWidth: 1,
    borderColor: Colors.TYP_MIDGRAY,
    borderRadius: Platform.select({
      ios: 33 / 2,
      android: 41 / 2,
    }),
    height: Platform.select({
      ios: 33,
      android: 41,
    }),
    paddingVertical: 6,
    paddingHorizontal: 12,
    justifyContent: 'space-evenly',
    alignItems: 'center',
  },
  timerText: {
    fontSize: 16,
    color: Colors.TEXT_BLACK,
    marginLeft: 5,
  },
  timerView: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
  },
  cancelText: {
    color: Colors.ICON_RED,
    fontSize: 16,
  },
});

export default class AttachAudio extends Component {
  constructor() {
    super();
    this.state = {
      recordTime: 0,
      filePath: '',
    };
  }

  componentWillMount() {
    this.recorder = null;
    this._reloadRecorder();
  }

  componentWillUnmount() {
    clearInterval(this._recordInterval);
    if (this.recorder) {
      this.recorder.destroy();
    }
  }

  _reloadRecorder() {
    if (this.recorder) {
      this.recorder.destroy();
    }

    this.recorder = new Recorder('newRecord.mp4', {
      bitrate: 256000,
      channels: 2,
      sampleRate: 44100,
      quality: 'max',
    })
      .prepare((err, fsPath) => {
        if (err) {
          // console.log('PREPARE RECORDER ERROR', err);
        } else {
          this.setState({ filePath: fsPath });
        }
      })
      .record((err) => {
        if (err) {
          // console.log('RECORDER ERROR', err);
          // this.setState({ recordTime: 0 });
          Actions.refresh({
            attachAudio: false,
          });
          return;
        }
        this._recordInterval = setInterval(() => {
          const { recordTime } = this.state;
          if (recordTime < 300) {
            this.setState({ recordTime: recordTime + 1 });
          } else {
            // clearInterval(this._recordInterval);
            this.cancelRecord();
          }
        }, 1000);
      });
  }

  sendRecord() {
    if (this.recorder && this.recorder.isRecording) {
      const { filePath } = this.state;
      this.recorder.stop((err) => {
        if (err) {
          // console.log('STOP RECORDER ERROR', err);
          if (this.recorder) {
            this.recorder.destroy();
          }
        }
        // clearInterval(this._recordInterval);
        // this.setState({ recordTime: 0 });
        Actions.refresh({
          attachAudio: false,
          dataToUpload: {
            uri: Platform.OS === 'ios' ? filePath : `file://${filePath}`,
          },
          imageCaption: 'Audio Message',
        });
      });
    }
  }

  cancelRecord() {
    if (this.recorder && this.recorder.isRecording) {
      this.recorder.stop((err) => {
        if (err) {
          // console.log('STOP RECORDER ERROR', err);
          if (this.recorder) {
            this.recorder.destroy();
          }
        }
        // clearInterval(this._recordInterval);
        // this.setState({ recordTime: 0 });
        Actions.refresh({
          attachAudio: false,
        });
      });
    }
  }

  render() {
    const { recordTime: recT } = this.state;
    const recM = (recT - (recT % 60)) / 60;
    const recS = recT % 60;
    const timerString = `${recM < 10 ? '0' : ''}${recM}:${recS < 10 ? '0' : ''}${recS}`;
    return (
      <View style={[customStyles.audioAttachContainer, styles.rowFlex]}>
        <View style={[styles.rowFlex, customStyles.composer]}>
          <View style={customStyles.timerView}>
            <View>
              <Icon
                name="checkbox-blank-circle"
                type="material-community"
                color={recT % 2 ? Colors.ICON_RED : Colors.TYP_GRAY}
                size={12}
              />
            </View>
            <Text numberOfLines={1} style={customStyles.timerText}>
              {timerString}
            </Text>
          </View>

          <TouchableOpacity onPress={() => this.cancelRecord()}>
            <Text numberOfLines={1} style={customStyles.cancelText}>
              Cancel
            </Text>
            {/* <FontAwesomeIcon
              name="trash-o"
              type="material-community"
              color={Colors.TYP_GRAY}
              size={22}
            /> */}
          </TouchableOpacity>
        </View>
        <TouchableOpacity onPress={() => this.sendRecord()}>
          <View style={[styles.chatDetailSendView, styles.alignJustifyCenter]}>
            <Icon name="send" color={Colors.TEXT_HEADER} size={28} />
          </View>
        </TouchableOpacity>
      </View>
    );
    // return (
    //   <View style={styles.audioAttachContainer}>
    //     <View style={styles.rowFlex}>
    //       <View style={styles.audioAttachTimerView}>
    //         <Text numberOfLines={1} ellipsizeMode="tail" style={styles.audioAttachTimerText}>{`${
    //           recM < 10 ? '0' : ''
    //         }${recM}:${recS < 10 ? '0' : ''}${recS}`}</Text>
    //       </View>
    //       <TouchableOpacity
    //         style={[styles.centerContainer, styles.margin5]}
    //         onPress={() => this.sendRecord()}
    //       >
    //         <View style={[styles.audioAttachiconButton, styles.audioAttachSendIconView]}>
    //           <Icon name="send" type="material-community" color={Colors.ICON_WHITE} size={24} />
    //         </View>
    //         <Text numberOfLines={1} ellipsizeMode="tail" style={styles.audioAttachText}>
    //           Send
    //         </Text>
    //       </TouchableOpacity>
    //       <TouchableOpacity
    //         style={[styles.centerContainer, styles.margin5]}
    //         onPress={() => this.cancelRecord()}
    //       >
    //         <View style={[styles.audioAttachiconButton, styles.audioAttachCancelIconView]}>
    //           <Icon name="close" type="material-community" color={Colors.ICON_WHITE} size={28} />
    //         </View>
    //         <Text numberOfLines={1} ellipsizeMode="tail" style={styles.audioAttachText}>
    //           Cancel
    //         </Text>
    //       </TouchableOpacity>
    //     </View>
    //   </View>
    // );
  }
}
