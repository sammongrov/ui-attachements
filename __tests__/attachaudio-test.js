import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Actions } from 'react-native-router-flux';
import AttachAudio from '../AttachAudio';

configure({ adapter: new Adapter() });

jest.mock('react-native-router-flux', () => ({
  Actions: {
    refresh: jest.fn(),
  },
}));

jest.mock('react-native-audio-toolkit', () => {
  function MockRecorder() {
    this.destroy = jest.fn();
    this.prepare = jest.fn((cb) => {
      cb(null, '//records/audio1.mp4');
      return this;
    });
    this.record = jest.fn((cb) => {
      cb(null);
      return this;
    });
    this.stop = jest.fn((cb) => {
      cb(null);
    });
    this.isRecording = true;
  }
  return { Recorder: MockRecorder };
});

const Audio = require('react-native-audio-toolkit');

const { Recorder } = Audio;

jest.useFakeTimers();

beforeEach(() => {
  jest.resetModules();
});

/* ------------------------- Snapshots ----------------------- */
it('AttachAudio renders correctly', () => {
  const tree = renderer.create(<AttachAudio />).toJSON();
  expect(tree).toMatchSnapshot();
});

/* ------------------- component methods --------------------- */

it('AttachAudio willMount', () => {
  const tree = shallow(<AttachAudio />);
  const instance = tree.instance();
  expect(instance.recorder).toBeInstanceOf(Recorder);
});

it('AttachAudio willUnmount and destroys recorder', () => {
  const tree = shallow(<AttachAudio />);
  const instance = tree.instance();
  tree.unmount();
  expect(clearInterval).toBeCalled();
  expect(instance.recorder.destroy).toBeCalled();
});

it('AttachAudio willUnmount - no recorder', () => {
  const tree = shallow(<AttachAudio />);
  const instance = tree.instance();
  instance.recorder = null;
  tree.unmount();
  expect(clearInterval).toBeCalled();
});

it('AttachAudio calls _reloadRecorder - old recorder is present', () => {
  setInterval.mockClear();
  const tree = shallow(<AttachAudio />);
  tree.setState({ recordTime: 101 });
  const instance = tree.instance();
  instance.recorder = new Recorder('testAudio.mp4', {});
  instance._reloadRecorder();
  expect(instance.recorder.prepare).toBeCalled();
  expect(instance.recorder.record).toBeCalled();
  expect(setInterval).toBeCalled();
  expect(tree.state().filePath).toMatch('//records/audio1.mp4');
  jest.advanceTimersByTime(1000);
  expect(tree.state().recordTime).toBe(103);
});

it('AttachAudio calls _reloadRecorder - record time > 300', () => {
  setInterval.mockClear();
  const tree = shallow(<AttachAudio />);
  tree.setState({ recordTime: 303 });
  const instance = tree.instance();
  instance.cancelRecord = jest.fn();
  instance._reloadRecorder();
  jest.advanceTimersByTime(1000);
  expect(tree.state().recordTime).toBe(303);
  expect(instance.cancelRecord).toBeCalled();
});

it('AttachAudio calls sendRecord - no recorder', () => {
  const tree = shallow(<AttachAudio />);
  const instance = tree.instance();
  instance.recorder = null;
  const result = instance.sendRecord();
  expect(result).toBeUndefined();
});

it('AttachAudio calls sendRecord - the recorder is not recording', () => {
  const tree = shallow(<AttachAudio />);
  const instance = tree.instance();
  instance.recorder.isRecording = false;
  const result = instance.sendRecord();
  expect(result).toBeUndefined();
});

it('AttachAudio calls sendRecord - ios', () => {
  jest.doMock('Platform', () => {
    const Platform = require.requireActual('Platform');
    Platform.OS = 'ios';
    return Platform;
  });
  Actions.refresh.mockClear();
  const filePath = '//records/audio1.mp4';
  const args = {
    attachAudio: false,
    dataToUpload: {
      uri: filePath,
    },
    imageCaption: 'Audio Message',
  };
  const tree = shallow(<AttachAudio />);
  const instance = tree.instance();
  instance.recorder.isRecording = true;
  instance.sendRecord();
  expect(Actions.refresh).toBeCalledWith(args);
});

it('AttachAudio calls sendRecord - android', () => {
  jest.doMock('Platform', () => {
    const Platform = require.requireActual('Platform');
    Platform.OS = 'android';
    return Platform;
  });
  Actions.refresh.mockClear();
  const filePath = '//records/audio1.mp4';
  const args = {
    attachAudio: false,
    dataToUpload: {
      uri: `file://${filePath}`,
    },
    imageCaption: 'Audio Message',
  };
  const tree = shallow(<AttachAudio />);
  const instance = tree.instance();
  instance.recorder.isRecording = true;
  instance.sendRecord();
  expect(Actions.refresh).toBeCalledWith(args);
});

it('AttachAudio calls sendRecord and destroy recorder', () => {
  jest.doMock('Platform', () => {
    const Platform = require.requireActual('Platform');
    Platform.OS = 'android';
    return Platform;
  });
  const error = new Error('cannot stop the recorder');
  const tree = shallow(<AttachAudio />);
  const instance = tree.instance();
  instance.recorder.destroy.mockClear();
  instance.recorder.isRecording = true;
  instance.recorder.stop = jest.fn((cb) => {
    cb(error);
  });
  instance.sendRecord();
  expect(instance.recorder.destroy).toBeCalled();
});

it('AttachAudio calls sendRecord and recorder is already destroyed', () => {
  jest.doMock('Platform', () => {
    const Platform = require.requireActual('Platform');
    Platform.OS = 'android';
    return Platform;
  });
  const error = new Error('cannot stop the recorder');
  const destroy = jest.fn();
  const tree = shallow(<AttachAudio />);
  const instance = tree.instance();
  instance.recorder.destroy = destroy;
  instance.recorder.isRecording = true;
  instance.recorder.stop = jest.fn((cb) => {
    instance.recorder = null;
    cb(error);
  });
  instance.sendRecord();
  expect(destroy).not.toBeCalled();
});

it('AttachAudio calls cancelRecord - no recorder', () => {
  const tree = shallow(<AttachAudio />);
  const instance = tree.instance();
  instance.recorder = null;
  const result = instance.cancelRecord();
  expect(result).toBeUndefined();
});

it('AttachAudio calls cancelRecord - the recorder is not recording', () => {
  const tree = shallow(<AttachAudio />);
  const instance = tree.instance();
  instance.recorder.isRecording = false;
  const result = instance.cancelRecord();
  expect(result).toBeUndefined();
});

it('AttachAudio calls cancelRecord', () => {
  Actions.refresh.mockClear();
  const tree = shallow(<AttachAudio />);
  const instance = tree.instance();
  instance.recorder.isRecording = true;
  instance.cancelRecord();
  expect(Actions.refresh).toBeCalledWith({ attachAudio: false });
});

it('AttachAudio calls cancelRecord and destroy recorder', () => {
  const error = new Error('cannot stop the recorder');
  const tree = shallow(<AttachAudio />);
  const instance = tree.instance();
  instance.recorder.destroy.mockClear();
  instance.recorder.isRecording = true;
  instance.recorder.stop = jest.fn((cb) => {
    cb(error);
  });
  instance.cancelRecord();
  expect(instance.recorder.destroy).toBeCalled();
});

it('AttachAudio calls cancelRecord and recorder is already destroyed', () => {
  const error = new Error('cannot stop the recorder');
  const destroy = jest.fn();
  const tree = shallow(<AttachAudio />);
  const instance = tree.instance();
  instance.recorder.destroy = destroy;
  instance.recorder.isRecording = true;
  instance.recorder.stop = jest.fn((cb) => {
    instance.recorder = null;
    cb(error);
  });
  instance.cancelRecord();
  expect(destroy).not.toBeCalled();
});

it('onPress of a cancel record button', () => {
  const tree = shallow(<AttachAudio />);
  tree.setState({ recordTime: 616 });
  const instance = tree.instance();
  instance.cancelRecord = jest.fn();
  const cancelButton = tree.find('TouchableOpacity').first();
  cancelButton.props().onPress();
  expect(instance.cancelRecord).toBeCalled();
});

it('onPress of a send record button', () => {
  const tree = shallow(<AttachAudio />);
  const instance = tree.instance();
  instance.sendRecord = jest.fn();
  const sendButton = tree.find('TouchableOpacity').last();
  sendButton.props().onPress();
  expect(instance.sendRecord).toBeCalled();
});

// call it the last
it('AttachAudio calls _reloadRecorder - an error in prepare', () => {
  jest.resetModules();
  jest.unmock('react-native-audio-toolkit');
  Audio.Recorder = function() {
    const error = new Error('recorder error');
    this.destroy = jest.fn();
    this.prepare = jest.fn((cb) => {
      cb(error, null);
      return this;
    });
    this.record = jest.fn((cb) => {
      cb(error);
      return this;
    });
    this.stop = jest.fn();
    this.isRecording = false;
  };
  setInterval.mockClear();
  const tree = shallow(<AttachAudio />);
  const instance = tree.instance();
  instance._reloadRecorder();
  expect(instance.recorder.prepare).toBeCalled();
  expect(tree.state().filePath).toBe('');
  expect(Actions.refresh).toBeCalledWith({ attachAudio: false });
  expect(setInterval).not.toBeCalled();
});
