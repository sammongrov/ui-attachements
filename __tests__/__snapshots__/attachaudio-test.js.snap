// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttachAudio renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "paddingBottom": 35,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        undefined,
        Object {
          "alignItems": "center",
          "borderColor": "#C7C7CC",
          "borderRadius": 16.5,
          "borderWidth": 1,
          "height": 33,
          "justifyContent": "space-evenly",
          "marginBottom": 5,
          "marginLeft": 10,
          "marginTop": 6,
          "paddingHorizontal": 12,
          "paddingVertical": 6,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View>
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#8E8E93",
                "fontSize": 12,
              },
              undefined,
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
          type="material-community"
        >
          
        </Text>
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Object {
            "color": "#000",
            "fontSize": 16,
            "marginLeft": 5,
          }
        }
      >
        00:00
      </Text>
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Object {
            "color": "#CF000F",
            "fontSize": 16,
          }
        }
      >
        Cancel
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          undefined,
          undefined,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#2E88FF",
              "fontSize": 28,
            },
            undefined,
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;
